modules/LocationUtils.js

```diff
@@ -65,6 +52,11 @@ export function createLocation(path, state, key, currentLocation) {
       location.pathname = '/';
     }
   }
+  
+  tinywarning(
+    location.pathname === encodeurl(location.pathname),
+    "location.pathname is not fully encoded, which may result in bugs."
+  );
```

Comment:
I have mixed feelings about libraries directly invoking the console API. In our application, we require all logs to be in JSON format (via a library such as Bunyan/Winston/Pino), so they can be easily queried at production scale.

I'm not sure what else we could do here though—other than provide a way to change the logging effect, defaulting to the native console API.

Reply:
`tinywarning` is a no-op in production. Would it be an issue for you in dev?

I feel like we need something to indicate that the `pathname` is expected to be fully encoded. The other suggestion, to throw, is interesting, but I thought might be a bit extreme.

Reply:
(Also, please continue to make changes yourself where you see fit. My intention here wasn't to completely hijack your PR.)

Reply:
Ah, if it's dev only, that would be fine.

If the `pathname` isn't encoded, what sort of problems will be occur? Sometimes it's good to fail early/loudly, but not sure if this is one of those or not.

Reply:
The primary issue that I see is with path matching hard-coded encoded characters. If an app expects `pathname`s to be encoded, its paths should be encoded too:

```jsx
<Route path="/Beyonc%C3%A9" component={Bey} />
```

A location that comes from the browser will have an encoded `pathname` and everything willb e fine. However, if the user pushes a non-encoded location (e.g. `/Beyoncé`), the route would fail to match.

Being confident that the `pathname` always has the same form was a big part of why the decoding was originally added (alongside decoded params).

With throwing, I worry that someone would push a non-encoded location, the app doesn't catch it (or, worse, swallows it), and the user is left in limbo.

```jsx
<Link to="/Beyoncé">Beyoncé</Link>
// *click* and nothing happens
```

If we warn, but let the navigation go through, encoded route `path`s would fail (comparing `/Beyoncé` to `/Beyonc%C3%A9`), but paths with params would still work (e.g. `/artist/:name` would work for either `/artist/Beyoncé` or `/artist/Beyonc%C3%A9`). That would enable partial functionality, which may be worse than no functionality if it leads to subtle bugs.

```jsx
<Route path="/artist/:name" />

// location.pathname = /artist/Beyoncé
match.params.name === "Beyoncé

// location.pathname = /artist/Beyonc%C3%A9
match.params.name === "Beyonc%C3%A9"
```

I'm not really sure what the best approach would be, so I'm kind of hoping someone shows up and says "x is better than y for reason z".

Reply:
Having typed that, I think that throwing would be more sound because it would not introduce bugs between encoded/non-encoded pathnames, but I would still like to hear input from others.

Reply:
Hmm, I think I agree.

> With throwing, I worry that someone would push a non-encoded location, the app doesn't catch it (or, worse, swallows it), and the user is left in limbo.

What if we didn't throw or warn? I guess this path just wouldn't match the route? In which case the user would presumably see a 404 (providing one had been setup).

On a side note, in your `params` example, I believe the actual behaviour should be:

```
<Route path="/artist/:name" />

// location.pathname = /artist/Beyonc%C3%A9
match.params.name === "Beyoncé"
```

`params` should be decoded (via `decodeURIComponent`), like they are in Express:

> decoding params using decodeURIComponent instead of the current decodeURI. This brings the behaviour inline with other frameworks such as Express, which automatically decodes the values in req.params using decodeURIComponent.

– https://github.com/ReactTraining/history/issues/505#issuecomment-375646353

Reply:
In a future React Router release it should decode, but right now there is nothing to support this. We can't safely call `decodeURIComponent` on the matched params until we are certain that the `pathname` is fully encoded, so the change will have to come in a new RR version. Maybe we could fiddle with `<Route>` props to support this in a minor (`<Route decode />`), but we'll cross that bridge when we get there.

Michael and I are going to chat about this in a few hours. This change _is_ going to be made, the big question for me is determining what to do when we don't get a fully encoded pathname.